tictactoe/src/main/java/models/TicTacToeService.java

```diff
@@ -0,0 +1,144 @@
+    initialiseGameBoard(grid);
+    displayGrid(grid);
+    List<Player> playerQueue;
+    playerQueue = gameboard.getPlayers();
+    for(int k=0;k<gameboard.getUserRowColumnInputs().size();k++) {
+      Player currentPlayer = playerQueue.get(0);
+      playerQueue.remove(0);
+      int inputRowNum = gameboard.getUserRowColumnInputs().get(k).getInputRowNum();
+      int inputColumnNum = gameboard.getUserRowColumnInputs().get(k).getInputColumnNum();
+      if(!isValidMove(grid,inputRowNum,inputColumnNum)) {
+        System.out.println("Invalid Move");
+        playerQueue.add(0,currentPlayer);
+      }
+      else {
+        grid[inputRowNum-1][inputColumnNum-1] = currentPlayer.getChosenCharacter()+" ";
+        displayGrid(grid);
+        if(isGameComplete(grid)) {
+          System.out.println("Game Over");
+          break;
+        }
+        if(rowCheck(grid,inputRowNum,inputColumnNum) || columnCheck(grid,inputRowNum,inputColumnNum)
+            || leftDiagonalCheck(grid,inputRowNum,inputColumnNum) || rightDiagonalCheck(grid,inputRowNum,inputColumnNum)) {
+          System.out.println(currentPlayer.getName()+" won the game");
+          break;
+        }
+        playerQueue.add(currentPlayer);
+      }
+    }
+  }
+
+  private void displayGrid(String[][] grid) {
+    for (String[] strings : grid) {
+      for (String string : strings) {
+        System.out.print(string);
+      }
+      System.out.println();
+    }
+  }
+
+  private boolean isValidMove(String [][]grid,int inputRowNum,int inputColumnNum) {
+    return grid[inputRowNum - 1][inputColumnNum - 1].equals("- ") && (inputRowNum >= 1
+        && inputRowNum <= grid.length)
+        && (inputColumnNum >= 1 && inputColumnNum <= grid.length);
+  }
+
+  private boolean columnCheck(String [][]grid,int inputRowNum,int inputColumnNum) {
+    for(int i=0;i<grid.length;i++) {
+      if(!grid[inputRowNum - 1][i].equals(grid[inputRowNum - 1][inputColumnNum - 1])) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private boolean rowCheck(String [][]grid,int inputRowNum,int inputColumnNum) {
+    for (String[] strings : grid) {
+      if (!strings[inputColumnNum - 1].equals(grid[inputRowNum - 1][inputColumnNum - 1])) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private boolean leftDiagonalCheck(String [][]grid,int inputRowNum,int inputColumnNum) {
+    int characterCounter=0;
+    for(int i=0;i<grid.length;i++) {
+      for(int j=0;j<grid.length;j++) {
+        if(i==j && grid[i][j].equals(grid[inputRowNum - 1][inputColumnNum - 1])) {
+          characterCounter++;
+        }
+      }
+    }
+    return characterCounter == grid.length;
+  }
+
+  private boolean rightDiagonalCheck(String [][]grid,int inputRowNum,int inputColumnNum) {
+    int characterCounter=0;
+    for(int i=grid.length-1;i>=0;i--) {
+      for(int j=grid.length-1;j>=0;j--) {
+        if(i==j && grid[i][j].equals(grid[inputRowNum - 1][inputColumnNum - 1])) {
```

Comment:
Is this tested? It would fail for certain cases. Do you need a nested loop?

Reply:
private boolean rightDiagonalCheck(String [][]grid,int inputRowNum,int inputColumnNum) {
    int characterCounter=0;
    int row=0;
    int column=grid.length-1;
    while(row<grid.length && column>=0) {
      if(row+column+1==grid.length && grid[row][column].equals(grid[inputRowNum - 1][inputColumnNum - 1])) {
          characterCounter++;
      }
      row++;
      column--;

    }
    return characterCounter == grid.length;
  }

Reply:
private boolean rightDiagonalCheck(String [][]grid,int inputRowNum,int inputColumnNum) {
    int characterCounter=0;
    int row=0;
    int column=grid.length-1;
    while(row<grid.length && column>=0) {
      if(row+column+1==grid.length && grid[row][column].equals(grid[inputRowNum - 1][inputColumnNum - 1])) {
          characterCounter++;
      }
      row++;
      column--;

    }
    return characterCounter == grid.length;
  }

Reply:
improved logic and avoided nested loops.